File: plugins/typescript/analyzer.js
#!/usr/bin/env node

/**
 * TypeScript/JavaScript Dependency Analyzer Plugin
 * 
 * This plugin uses the TypeScript Compiler API to perform accurate AST analysis
 * and detect dependencies between files for intelligent PR partitioning.
 * 
 * Input: JSON via stdin with changed files and project context
 * Output: JSON to stdout with dependency relationships
 */

const fs = require('fs');
const path = require('path');

// Try to require TypeScript - gracefully handle if not installed
let ts;
try {
    ts = require('typescript');
} catch (err) {
    console.error(JSON.stringify({
        dependencies: [],
        errors: ['TypeScript not installed. Run: npm install typescript'],
        metadata: {
            filesAnalyzed: 0,
            analysisTime: "0ms",
            pluginName: "typescript-analyzer",
            pluginVersion: "1.0.0"
        }
    }));
    process.exit(1);
}

// Read input from stdin
let inputData = '';
process.stdin.setEncoding('utf8');

process.stdin.on('readable', () => {
    let chunk;
    while ((chunk = process.stdin.read()) !== null) {
        inputData += chunk;
    }
});

process.stdin.on('end', () => {
    const startTime = Date.now();
    
    try {
        const input = JSON.parse(inputData);
        const result = analyzeTypeScriptDependencies(input, startTime);
        console.log(JSON.stringify(result, null, 2));
    } catch (error) {
        console.error(JSON.stringify({
            dependencies: [],
            errors: [error.message],
            metadata: {
                filesAnalyzed: 0,
                analysisTime: `${Date.now() - startTime}ms`,
                pluginName: "typescript-analyzer",
                pluginVersion: "1.0.0"
            }
        }));
        process.exit(1);
    }
});

/**
 * Main analysis function
 */
function analyzeTypeScriptDependencies(input, startTime) {
    const { changedFiles, projectFiles, projectRoot } = input;
    const allFiles = [...changedFiles, ...projectFiles];
    
    // Filter TypeScript/JavaScript files
    const tsFiles = allFiles.filter(file => 
        file.path.match(/\.(ts|tsx|js|jsx)$/i)
    );
    
    if (tsFiles.length === 0) {
        return {
            dependencies: [],
            metadata: {
                filesAnalyzed: 0,
                analysisTime: `${Date.now() - startTime}ms`,
                pluginName: "typescript-analyzer",
                pluginVersion: "1.0.0"
            },
            errors: []
        };
    }
    
    // Create TypeScript program for analysis
    const { program, sourceFiles } = createTypeScriptProgram(tsFiles, projectRoot);
    
    // Analyze dependencies
    const dependencies = [];
    const errors = [];
    
    try {
        // Analyze each changed file for dependencies
        for (const file of changedFiles) {
            if (!file.path.match(/\.(ts|tsx|js|jsx)$/i)) {
                continue;
            }
            
            const sourceFile = sourceFiles.get(file.path);
            if (sourceFile) {
                const fileDeps = analyzeSourceFile(sourceFile, program, allFiles);
                dependencies.push(...fileDeps);
            }
        }
        
        // Remove duplicates
        const uniqueDependencies = removeDuplicateDependencies(dependencies);
        
        return {
            dependencies: uniqueDependencies,
            metadata: {
                filesAnalyzed: tsFiles.length,
                analysisTime: `${Date.now() - startTime}ms`,
                pluginName: "typescript-analyzer",
                pluginVersion: "1.0.0"
            },
            errors: errors
        };
        
    } catch (error) {
        errors.push(`Analysis error: ${error.message}`);
        
        return {
            dependencies: dependencies,
            metadata: {
                filesAnalyzed: tsFiles.length,
                analysisTime: `${Date.now() - startTime}ms`,
                pluginName: "typescript-analyzer",
                pluginVersion: "1.0.0"
            },
            errors: errors
        };
    }
}

/**
 * Create TypeScript program for analysis
 */
function createTypeScriptProgram(files, projectRoot) {
    // Create in-memory file system for TypeScript
    const fileMap = new Map();
    const sourceFiles = new Map();
    
    // Add files to in-memory system
    for (const file of files) {
        const fullPath = path.resolve(projectRoot, file.path);
        fileMap.set(fullPath, file.content);
    }
    
    // TypeScript compiler options
    const compilerOptions = {
        target: ts.ScriptTarget.ES2020,
        module: ts.ModuleKind.ESNext,
        moduleResolution: ts.ModuleResolutionKind.NodeJs,
        allowJs: true,
        checkJs: false,
        jsx: ts.JsxEmit.React,
        esModuleInterop: true,
        allowSyntheticDefaultImports: true,
        skipLibCheck: true,
        noResolve: false
    };
    
    // Custom compiler host
    const compilerHost = {
        getSourceFile: (fileName, languageVersion) => {
            const content = fileMap.get(fileName);
            if (content !== undefined) {
                const sourceFile = ts.createSourceFile(fileName, content, languageVersion, true);
                sourceFiles.set(path.relative(projectRoot, fileName).replace(/\\/g, '/'), sourceFile);
                return sourceFile;
            }
            
            // Try to read from file system for external dependencies
            try {
                if (fs.existsSync(fileName)) {
                    const content = fs.readFileSync(fileName, 'utf8');
                    return ts.createSourceFile(fileName, content, languageVersion, true);
                }
            } catch (err) {
                // Ignore file system errors
            }
            
            return undefined;
        },
        
        writeFile: () => {}, // No-op
        getCurrentDirectory: () => projectRoot,
        getDirectories: () => [],
        fileExists: (fileName) => fileMap.has(fileName) || fs.existsSync(fileName),
        readFile: (fileName) => fileMap.get(fileName) || (fs.existsSync(fileName) ? fs.readFileSync(fileName, 'utf8') : undefined),
        getCanonicalFileName: (fileName) => fileName,
        useCaseSensitiveFileNames: () => true,
        getNewLine: () => '\n'
    };
    
    // Create program
    const program = ts.createProgram(
        Array.from(fileMap.keys()),
        compilerOptions,
        compilerHost
    );
    
    return { program, sourceFiles };
}

/**
 * Analyze a source file for dependencies
 */
function analyzeSourceFile(sourceFile, program, allFiles) {
    const dependencies = [];
    const fileName = sourceFile.fileName;
    const checker = program.getTypeChecker();
    
    // Get relative path for consistent reporting
    const relativePath = path.relative(process.cwd(), fileName).replace(/\\/g, '/');
    
    // Visit all nodes in the AST
    function visit(node) {
        // Import declarations: import ... from "module"
        if (ts.isImportDeclaration(node)) {
            const dependency = analyzeImportDeclaration(node, relativePath, allFiles);
            if (dependency) {
                dependencies.push(dependency);
            }
        }
        
        // Export declarations: export ... from "module"
        else if (ts.isExportDeclaration(node)) {
            const dependency = analyzeExportDeclaration(node, relativePath, allFiles);
            if (dependency) {
                dependencies.push(dependency);
            }
        }
        
        // Call expressions: require("module"), import("module")
        else if (ts.isCallExpression(node)) {
            const dependency = analyzeCallExpression(node, relativePath, allFiles, checker);
            if (dependency) {
                dependencies.push(dependency);
            }
        }
        
        // Property access expressions for deeper analysis
        else if (ts.isPropertyAccessExpression(node)) {
            const dependency = analyzePropertyAccess(node, relativePath, allFiles, checker);
            if (dependency) {
                dependencies.push(dependency);
            }
        }
        
        // Continue traversing
        ts.forEachChild(node, visit);
    }
    
    visit(sourceFile);
    return dependencies;
}

/**
 * Analyze import declarations
 */
function analyzeImportDeclaration(node, fromFile, allFiles) {
    if (!node.moduleSpecifier || !ts.isStringLiteral(node.moduleSpecifier)) {
        return null;
    }
    
    const modulePath = node.moduleSpecifier.text;
    const resolvedPath = resolveModulePath(modulePath, fromFile, allFiles);
    
    if (!resolvedPath) {
        return null;
    }
    
    // Determine import type and strength
    let importType = 'import';
    let strength = 'CRITICAL'; // Imports are critical dependencies
    
    // Check if it's a type-only import
    if (node.importClause && node.importClause.isTypeOnly) {
        importType = 'type-import';
        strength = 'MODERATE'; // Type imports are moderate dependencies
    }
    
    return {
        from: fromFile,
        to: resolvedPath,
        type: importType,
        strength: strength,
        line: getLineNumber(node),
        context: getNodeText(node)
    };
}

/**
 * Analyze export declarations
 */
function analyzeExportDeclaration(node, fromFile, allFiles) {
    if (!node.moduleSpecifier || !ts.isStringLiteral(node.moduleSpecifier)) {
        return null;
    }
    
    const modulePath = node.moduleSpecifier.text;
    const resolvedPath = resolveModulePath(modulePath, fromFile, allFiles);
    
    if (!resolvedPath) {
        return null;
    }
    
    return {
        from: fromFile,
        to: resolvedPath,
        type: 'export',
        strength: 'CRITICAL', // Re-exports are critical
        line: getLineNumber(node),
        context: getNodeText(node)
    };
}

/**
 * Analyze call expressions (require, dynamic import)
 */
function analyzeCallExpression(node, fromFile, allFiles, checker) {
    // require() calls
    if (ts.isIdentifier(node.expression) && node.expression.text === 'require') {
        if (node.arguments.length > 0 && ts.isStringLiteral(node.arguments[0])) {
            const modulePath = node.arguments[0].text;
            const resolvedPath = resolveModulePath(modulePath, fromFile, allFiles);
            
            if (resolvedPath) {
                return {
                    from: fromFile,
                    to: resolvedPath,
                    type: 'require',
                    strength: 'STRONG', // require() is strong dependency
                    line: getLineNumber(node),
                    context: getNodeText(node)
                };
            }
        }
    }
    
    // Dynamic import() calls
    if (node.expression.kind === ts.SyntaxKind.ImportKeyword) {
        if (node.arguments.length > 0 && ts.isStringLiteral(node.arguments[0])) {
            const modulePath = node.arguments[0].text;
            const resolvedPath = resolveModulePath(modulePath, fromFile, allFiles);
            
            if (resolvedPath) {
                return {
                    from: fromFile,
                    to: resolvedPath,
                    type: 'dynamic-import',
                    strength: 'MODERATE', // Dynamic imports are moderate
                    line: getLineNumber(node),
                    context: getNodeText(node)
                };
            }
        }
    }
    
    return null;
}

/**
 * Analyze property access for method calls and deeper dependencies
 */
function analyzePropertyAccess(node, fromFile, allFiles, checker) {
    // This could be expanded to detect more complex dependencies
    // For now, we'll keep it simple and return null
    // Future enhancement: detect method calls across modules
    return null;
}

/**
 * Resolve module path to actual file path
 */
function resolveModulePath(modulePath, fromFile, allFiles) {
    // Skip external modules (no relative path)
    if (!modulePath.startsWith('.')) {
        return null;
    }
    
    // Resolve relative path
    const fromDir = path.dirname(fromFile);
    let resolved = path.join(fromDir, modulePath);
    resolved = path.normalize(resolved).replace(/\\/g, '/');
    
    // Create a map of available files for quick lookup
    const fileMap = new Map();
    for (const file of allFiles) {
        fileMap.set(file.path, file);
        
        // Also map without extension for easier lookup
        const withoutExt = file.path.replace(/\.(ts|tsx|js|jsx)$/, '');
        fileMap.set(withoutExt, file);
    }
    
    // Try different extensions and variations
    const candidates = [
        resolved,
        resolved + '.ts',
        resolved + '.tsx',
        resolved + '.js',
        resolved + '.jsx',
        resolved + '/index.ts',
        resolved + '/index.tsx',
        resolved + '/index.js',
        resolved + '/index.jsx'
    ];
    
    for (const candidate of candidates) {
        if (fileMap.has(candidate)) {
            return candidate;
        }
    }
    
    return null;
}

/**
 * Get line number of a node
 */
function getLineNumber(node) {
    if (node.getSourceFile) {
        const sourceFile = node.getSourceFile();
        const lineAndChar = sourceFile.getLineAndCharacterOfPosition(node.getStart());
        return lineAndChar.line + 1;
    }
    return 0;
}

/**
 * Get text content of a node
 */
function getNodeText(node) {
    if (node.getSourceFile) {
        return node.getSourceFile().text.substring(node.getStart(), node.getEnd()).trim();
    }
    return '';
}

/**
 * Remove duplicate dependencies
 */
function removeDuplicateDependencies(dependencies) {
    const seen = new Set();
    const unique = [];
    
    for (const dep of dependencies) {
        const key = `${dep.from}:${dep.to}:${dep.type}`;
        if (!seen.has(key)) {
            seen.add(key);
            unique.push(dep);
        }
    }
    
    return unique;
}

File: plugins/python/analyzer.py
#!/usr/bin/env python3

"""
Python Dependency Analyzer Plugin

This plugin uses Python's AST module to analyze import statements
and detect dependencies between Python files for intelligent PR partitioning.

Input: JSON via stdin with changed files and project context
Output: JSON to stdout with dependency relationships
"""

import json
import sys
import ast
import os
import time
from pathlib import Path
from typing import List, Dict, Any, Optional

class PythonDependencyAnalyzer:
    def __init__(self):
        self.start_time = time.time()
        
    def analyze(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Main analysis function"""
        changed_files = input_data.get('changedFiles', [])
        project_files = input_data.get('projectFiles', [])
        project_root = input_data.get('projectRoot', os.getcwd())
        
        all_files = changed_files + project_files
        
        # Filter Python files
        python_files = [f for f in all_files if self._is_python_file(f['path'])]
        
        if not python_files:
            return self._create_result([], 0, [])
        
        # Create file mapping for quick lookup
        file_map = {f['path']: f for f in all_files}
        
        # Analyze dependencies
        dependencies = []
        errors = []
        
        try:
            for file_info in changed_files:
                if not self._is_python_file(file_info['path']):
                    continue
                    
                file_deps = self._analyze_file(file_info, file_map, project_root)
                dependencies.extend(file_deps)
                
        except Exception as e:
            errors.append(f"Analysis error: {str(e)}")
        
        # Remove duplicates
        unique_deps = self._remove_duplicates(dependencies)
        
        return self._create_result(unique_deps, len(python_files), errors)
    
    def _is_python_file(self, file_path: str) -> bool:
        """Check if file is a Python file"""
        return file_path.endswith(('.py', '.pyi'))
    
    def _analyze_file(self, file_info: Dict[str, Any], file_map: Dict[str, Any], project_root: str) -> List[Dict[str, Any]]:
        """Analyze a single Python file for dependencies"""
        dependencies = []
        
        try:
            # Parse the Python file using AST
            tree = ast.parse(file_info['content'], filename=file_info['path'])
            
            # Visit all nodes to find import statements
            for node in ast.walk(tree):
                dep = None
                
                if isinstance(node, ast.Import):
                    # Handle: import module, import module.submodule
                    dep = self._analyze_import(node, file_info['path'], file_map, project_root)
                    
                elif isinstance(node, ast.ImportFrom):
                    # Handle: from module import something
                    dep = self._analyze_import_from(node, file_info['path'], file_map, project_root)
                
                if dep:
                    dependencies.append(dep)
                    
        except SyntaxError as e:
            # Skip files with syntax errors (might be Python 2, etc.)
            pass
        except Exception as e:
            # Log other errors but continue
            pass
            
        return dependencies
    
    def _analyze_import(self, node: ast.Import, from_file: str, file_map: Dict[str, Any], project_root: str) -> Optional[Dict[str, Any]]:
        """Analyze import statements"""
        for alias in node.names:
            module_path = self._resolve_module_path(alias.name, from_file, file_map, project_root)
            if module_path:
                return {
                    'from': from_file,
                    'to': module_path,
                    'type': 'import',
                    'strength': 'CRITICAL',
                    'line': node.lineno,
                    'context': f"import {alias.name}"
                }
        return None
    
    def _analyze_import_from(self, node: ast.ImportFrom, from_file: str, file_map: Dict[str, Any], project_root: str) -> Optional[Dict[str, Any]]:
        """Analyze from...import statements"""
        if not node.module:
            # Relative import without module (from . import something)
            return None
            
        module_path = self._resolve_module_path(node.module, from_file, file_map, project_root, node.level)
        if module_path:
            imported_items = [alias.name for alias in node.names]
            return {
                'from': from_file,
                'to': module_path,
                'type': 'from-import',
                'strength': 'STRONG',
                'line': node.lineno,
                'context': f"from {node.module} import {', '.join(imported_items)}"
            }
        return None
    
    def _resolve_module_path(self, module_name: str, from_file: str, file_map: Dict[str, Any], project_root: str, level: int = 0) -> Optional[str]:
        """Resolve module name to actual file path"""
        # Skip standard library and external modules
        if self._is_standard_library(module_name):
            return None
            
        # Handle relative imports
        if level > 0:
            # Relative import
            from_dir = Path(from_file).parent
            for _ in range(level - 1):
                from_dir = from_dir.parent
            base_path = from_dir / module_name.replace('.', '/')
        else:
            # Absolute import
            base_path = Path(project_root) / module_name.replace('.', '/')
        
        # Try different file extensions and __init__.py
        candidates = [
            str(base_path) + '.py',
            str(base_path / '__init__.py'),
            str(base_path) + '.pyi'
        ]
        
        for candidate in candidates:
            # Convert to relative path from project root
            try:
                rel_path = os.path.relpath(candidate, project_root).replace('\\', '/')
                if rel_path in file_map:
                    return rel_path
            except ValueError:
                # Path is not relative to project root
                continue
                
        return None
    
    def _is_standard_library(self, module_name: str) -> bool:
        """Check if module is part of Python standard library"""
        stdlib_modules = {
            'os', 'sys', 'json', 'time', 'datetime', 'pathlib', 'ast', 're',
            'collections', 'itertools', 'functools', 'operator', 'typing',
            'math', 'random', 'string', 'io', 'logging', 'unittest', 'urllib',
            'http', 'email', 'xml', 'html', 'csv', 'sqlite3', 'threading',
            'multiprocessing', 'subprocess', 'socket', 'ssl', 'hashlib',
            'base64', 'pickle', 'copy', 'tempfile', 'shutil', 'glob'
        }
        
        # Check if it's a known stdlib module or starts with stdlib prefix
        root_module = module_name.split('.')[0]
        return root_module in stdlib_modules
    
    def _remove_duplicates(self, dependencies: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Remove duplicate dependencies"""
        seen = set()
        unique = []
        
        for dep in dependencies:
            key = f"{dep['from']}:{dep['to']}:{dep['type']}"
            if key not in seen:
                seen.add(key)
                unique.append(dep)
                
        return unique
    
    def _create_result(self, dependencies: List[Dict[str, Any]], files_analyzed: int, errors: List[str]) -> Dict[str, Any]:
        """Create the result JSON"""
        analysis_time = int((time.time() - self.start_time) * 1000)
        
        return {
            'dependencies': dependencies,
            'metadata': {
                'filesAnalyzed': files_analyzed,
                'analysisTime': f"{analysis_time}ms",
                'pluginName': 'python-analyzer',
                'pluginVersion': '1.0.0'
            },
            'errors': errors
        }

def main():
    """Main entry point"""
    try:
        # Read input from stdin
        input_data = json.load(sys.stdin)
        
        # Create analyzer and run analysis
        analyzer = PythonDependencyAnalyzer()
        result = analyzer.analyze(input_data)
        
        # Output result as JSON
        json.dump(result, sys.stdout, indent=2)
        
    except Exception as e:
        error_result = {
            'dependencies': [],
            'errors': [str(e)],
            'metadata': {
                'filesAnalyzed': 0,
                'analysisTime': '0ms',
                'pluginName': 'python-analyzer',
                'pluginVersion': '1.0.0'
            }
        }
        json.dump(error_result, sys.stdout, indent=2)
        sys.exit(1)

if __name__ == '__main__':
    main() 